File: compilor/LL0.cpp
#include <bits/stdc++.h>
using namespace std;

struct Item {
    string lhs, rhs;
    int dotPos;

    Item(string l, string r, int d) : lhs(l), rhs(r), dotPos(d) {}

    bool operator==(const Item &other) const {
        return lhs == other.lhs && rhs == other.rhs && dotPos == other.dotPos;
    }

    bool operator<(const Item &other) const {
        return tie(lhs, rhs, dotPos) < tie(other.lhs, other.rhs, other.dotPos);
    }

    string toString() const {
        return lhs + " -> " + rhs.substr(0, dotPos) + "." + rhs.substr(dotPos);
    }
};

struct State {
    set<Item> items;
    map<char, int> transitions;

    bool operator==(const State &other) const {
        return items == other.items;
    }
};

vector<string> grammar;
vector<State> states;

void readGrammar(const string &filename) {
    ifstream file(filename);
    string line;
    while (getline(file, line)) {
        grammar.push_back(line);
    }
    file.close();
}

set<Item> closure(set<Item> items) {
    set<Item> closureSet = items;
    bool changed;
    do {
        changed = false;
        set<Item> newItems;
        for (const Item &item : closureSet) {
            if (item.dotPos < item.rhs.length()) {
                char symbol = item.rhs[item.dotPos];
                for (const string &rule : grammar) {
                    if (rule[0] == symbol) {
                        string rhs = rule.substr(3);
                        Item newItem(string(1, symbol), rhs, 0);
                        if (closureSet.find(newItem) == closureSet.end()) {
                            newItems.insert(newItem);
                            changed = true;
                        }
                    }
                }
            }
        }
        closureSet.insert(newItems.begin(), newItems.end());
    } while (changed);
    return closureSet;
}

set<Item> gotoFunction(const set<Item> &state, char symbol) {
    set<Item> nextState;
    for (const Item &item : state) {
        if (item.dotPos < item.rhs.length() && item.rhs[item.dotPos] == symbol) {
            nextState.insert(Item(item.lhs, item.rhs, item.dotPos + 1));
        }
    }
    return closure(nextState);
}

void constructLR0() {
    set<Item> startState = closure({Item("S'", "S", 0)});
    states.push_back({startState, {}});

    queue<int> queue;
    queue.push(0);

    while (!queue.empty()) {
        int currentIndex = queue.front();
        queue.pop();

        set<char> symbols;
        for (const Item &item : states[currentIndex].items) {
            if (item.dotPos < item.rhs.length()) {
                symbols.insert(item.rhs[item.dotPos]);
            }
        }

        for (char symbol : symbols) {
            set<Item> nextStateItems = gotoFunction(states[currentIndex].items, symbol);
            if (!nextStateItems.empty()) {
                State nextState = {nextStateItems, {}};
                auto it = find(states.begin(), states.end(), nextState);
                int nextIndex;
                if (it == states.end()) {
                    states.push_back(nextState);
                    nextIndex = states.size() - 1;
                    queue.push(nextIndex);
                } else {
                    nextIndex = distance(states.begin(), it);
                }
                states[currentIndex].transitions[symbol] = nextIndex;
            }
        }
    }
}

void printDFA() {
    for (size_t i = 0; i < states.size(); i++) {
        cout << "State " << i << ":\n";
        for (const Item &item : states[i].items) {
            cout << "  " << item.toString() << "\n";
        }
        for (const auto &[symbol, nextIndex] : states[i].transitions) {
            cout << "  On '" << symbol << "' -> State " << nextIndex << "\n";
        }
        cout << "\n";
    }
}

int main() {
    readGrammar("C:\\Users\\Mayank bharti\\Downloads\\c++\\compilor\\demoLR0.txt");
    constructLR0();
    printDFA();
    return 0;
}

File: compilor/LL0.py
import queue

class Item:
    def __init__(self, lhs, rhs, dot_pos):
        self.lhs = lhs
        self.rhs = rhs
        self.dot_pos = dot_pos

    def __eq__(self, other):
        return (self.lhs, self.rhs, self.dot_pos) == (other.lhs, other.rhs, other.dot_pos)

    def __hash__(self):
        return hash((self.lhs, self.rhs, self.dot_pos))

    def __str__(self):
        return f"{self.lhs} -> {self.rhs[:self.dot_pos]}.{self.rhs[self.dot_pos:]}"

class State:
    def __init__(self, items):
        self.items = set(items)
        self.transitions = {}

    def __eq__(self, other):
        return self.items == other.items

    def __hash__(self):
        return hash(tuple(sorted(self.items)))

grammar = []

def read_grammar(filename):
    with open(filename, "r") as file:
        for line in file:
            grammar.append(line.strip())

def closure(items):
    closure_set = set(items)
    changed = True
    while changed:
        changed = False
        new_items = set()
        for item in closure_set:
            if item.dot_pos < len(item.rhs):
                symbol = item.rhs[item.dot_pos]
                for rule in grammar:
                    if rule[0] == symbol:
                        new_item = Item(symbol, rule[3:], 0)
                        if new_item not in closure_set:
                            new_items.add(new_item)
                            changed = True
        closure_set.update(new_items)
    return closure_set

def goto(state, symbol):
    next_items = {Item(item.lhs, item.rhs, item.dot_pos + 1) for item in state if item.dot_pos < len(item.rhs) and item.rhs[item.dot_pos] == symbol}
    return closure(next_items)

states = []

def construct_lr0():
    start_state = closure({Item("S'", "S", 0)})
    states.append(State(start_state))
    q = queue.Queue()
    q.put(states[0])

    while not q.empty():
        current_state = q.get()
        symbols = {item.rhs[item.dot_pos] for item in current_state.items if item.dot_pos < len(item.rhs)}

        for symbol in symbols:
            next_items = goto(current_state.items, symbol)
            if next_items:
                new_state = State(next_items)
                if new_state not in states:
                    states.append(new_state)
                    q.put(new_state)
                current_state.transitions[symbol] = states.index(new_state)

def print_dfa():
    for i, state in enumerate(states):
        print(f"State {i}:")
        for item in state.items:
            print(f"  {item}")
        for symbol, target in state.transitions.items():
            print(f"  On '{symbol}' -> State {target}")
        print()

read_grammar("C:\\Users\\Mayank bharti\\Downloads\\c++\\compilor\\demoLR0.txt")
construct_lr0()
print_dfa()